test/metabase/lib/test_util/generators.cljc

```diff
@@ -167,6 +162,38 @@
           (is (= (count after-breakouts)
                  (count before-breakouts))))))))
 
+;; Filters =======================================================================================
+(add-step {:kind :filter})
+
+(defmethod next-steps* :filter [query _filter]
+  (let [stage-number (choose-stage query)]
+    [:filter stage-number (gen.filters/gen-filter (lib/filterable-columns query stage-number))]))
+
+(defmethod run-step* :filter [query [_filter stage-number filter-clause]]
+  (lib/filter query stage-number filter-clause))
+
+(defmethod before-and-after :filter [before after [_filter stage-number filter-clause]]
+  (let [before-filters (lib/filters before stage-number)
+        after-filters  (lib/filters after  stage-number)]
+    (testing (str ":filter stage " stage-number
+                  "\n\nBefore query\n" (u/pprint-to-str before)
+                  "\n\nAfter query\n" (u/pprint-to-str after)
+                  "\n\nwith filter clause\n" (u/pprint-to-str filter-clause)
+                  "\n")
+      (if (some #(lib.equality/= % filter-clause) before-filters)
```

Comment:
The namespace separator in `lib.equality/=` always tricks my brain into thinking this is a call to `/=` i.e `!=`, and I'm momentarily confused why it's doing the opposite of what I expect 😂